lib/rouge/lexers/openedge.rb

```diff
@@ -0,0 +1,399 @@
+SUB-MINIMUM SUBSCRIBE SUBST SUBSTITUTE SUBSTR SUBSTRING SUB-TOTAL SUBTYPE SUM 
+SUM-MAX SUPER SUPER-PROCEDURES SUPPRESS-NAMESPACE-PROCESSING SUPPRESS-W 
+SUPPRESS-WARNINGS SYMMETRIC-ENCRYPTION-ALGORITHM SYMMETRIC-ENCRYPTION-IV 
+SYMMETRIC-ENCRYPTION-KEY SYMMETRIC-SUPPORT SYSTEM-ALERT SYSTEM-ALERT-BOXES 
+SYSTEM-DIALOG SYSTEM-HELP SYSTEM-ID TABLE TABLE-HANDLE TABLE-NUMBER TABLE-SCAN
+TAB-POSITION TAB-STOP TARGET TARGET-PROCEDURE TEMP-DIR TEMP-DIRECTORY TEMP-TABLE
+TEMP-TABLE-PREPARE TERM TERMINAL TERMINATETEXT TEXT TEXT-CURSOR TEXT-SEG-GROW
+TEXT-SELECTED THEN THIS-OBJECT THIS-PROCEDURE THREAD-SAFE THREE-D THROUGH THROW
+THRU TIC-MARKS TIME TIME-SOURCE TITLE TITLE-BGC TITLE-BGCOLOR TITLE-DC
+TITLE-DCOLORTITLE-FGC TITLE-FGCOLOR TITLE-FO TITLE-FONT TO TODAY TOGGLE-BOX
+TOOLTIP TOOLTIPS TOPIC TOP-NAV-QUERY TOP-ONLY TO-ROWID TOTAL TRAILING TRANS
+TRANSACTION TRANSACTION-MODE TRANS-INIT-PROCEDURE TRANSPARENT TRIGGER TRIGGERS
+TRIM TRUE TRUNC TRUNCATE TYPE TYPE-OF UNBOX UNBUFF UNBUFFERED UNDERL UNDERLINE
+UNDO UNFORM UNFORMATTED UNION UNIQUE UNIQUE-ID UNIQUE-MATCH UNIX UNLESS-HIDDEN
+UNLOAD UNSIGNED-LONG UNSUBSCRIBE UP UPDATE UPDATE-ATTRIBUTE URL URL-DECODE
+URL-ENCODE URL-PASSWORD URL-USERID USE USE-DICT-EXPS USE-FILENAME USE-INDEX USER
+USE-REVVIDEO USERID USER-ID USE-TEXT USE-UNDERLINE USE-WIDGET-POOL USING
+V6DISPLAY V6FRAME VALIDATE VALIDATE-EXPRESSION VALIDATE-MESSAGE VALIDATE-SEAL
+VALIDATION-ENABLED VALID-EVENT VALID-HANDLE VALID-OBJECT VALUE VALUE-CHANGED
+VALUES VAR VARIABLE VERBOSE VERSION VERT VERTICAL VIEW VIEW-AS
+VIEW-FIRST-COLUMN-ON-REOPEN VIRTUAL-HEIGHT VIRTUAL-HEIGHT-CHARS VIRTUAL-HEIGHT-P
+VIRTUAL-HEIGHT-PIXELS VIRTUAL-WIDTH VIRTUAL-WIDTH-CHARS VIRTUAL-WIDTH-P
+VIRTUAL-WIDTH-PIXELS VISIBLE VOID WAIT WAIT-FOR WARNING WEB-CONTEXT WEEKDAY WHEN
+WHERE WHILE WIDGET WIDGET-E WIDGET-ENTER WIDGET-ID WIDGET-L WIDGET-LEAVE
+WIDGET-POOL WIDTH WIDTH WIDTH-CHARS WIDTH-P WIDTH-PIXELS WINDOW WINDOW-MAXIM
+WINDOW-MAXIMIZED WINDOW-MINIM WINDOW-MINIMIZED WINDOW-NAME WINDOW-NORMAL
+WINDOW-STA WINDOW-STATE WINDOW-SYSTEM WITH WORD-INDEX WORD-WRAP
+WORK-AREA-HEIGHT-PIXELS WORK-AREA-WIDTH-PIXELS WORK-AREA-X WORK-AREA-Y WORKFILE
+WORK-TAB WORK-TABLE WRITE WRITE-CDATA WRITE-CHARACTERS WRITE-COMMENT
+WRITE-DATA-ELEMENT WRITE-EMPTY-ELEMENT WRITE-ENTITY-REF WRITE-EXTERNAL-DTD
+WRITE-FRAGMENT WRITE-JSON WRITE-MESSAGE WRITE-PROCESSING-INSTRUCTION
+WRITE-STATUS WRITE-XML WRITE-XMLSCHEMA X XCODE XML-DATA-TYPE
+XML-ENTITY-EXPANSION-LIMIT XML-NODE-TYPE XML-SCHEMA-PATH
+XML-STRICT-ENTITY-RESOLUTION XML-SUPPRESS-NAMESPACE-PROCESSING X-OF XREF
+XREF-XML Y YEAR YEAR-OFFSET YES YES-NO YES-NO-CANCEL Y-OF
+        )
+      end
+
+      def self.keywords_type
+        @keywords_type ||= Set.new %w(
+BLOB CHARACTER CHAR CLOB COM-HANDLE DATE DATETIME DATETIME-TZ DECIMAL DEC 
+HANDLE INT64 INTEGER INT LOGICAL LONGCHAR MEMPTR RAW RECID ROWID WIDGET-HANDLE          
+        )
+      end
+
+      start { push :bol }
```

Comment:
I think there might be some terminology confusion here. `bol` is, notwithstanding the acronym, meant to represent the beginning of the file. It's used with `start` to allow for an initial state that will apply to the beginning of a file but that, once popped off, will no longer form part of the lexing rules.

You're using it in the definition of `:whitespace` which is itself used in a number of states. You also have `:expr_bol` which, despite being introduced with a comment that says it's the same as `:bol` but without labels, includes as its first and only express rule the creation of a label.

Perhaps these are some artefacts from the development process? I'm not familiar with OpenEdge ABL. Is it possible to explain simply what can uniquely come at the beginning of a file?